include/exadg/operators/solution_projection_between_triangulations.cpp

```diff
@@ -0,0 +1,414 @@
+        InverseMassType::MatrixfreeOperator :
+        InverseMassType::ElementwiseKrylovSolver;
+
+    preconditioner = std::make_shared<InverseMassPreconditioner<dim, n_components, Number>>(
+      target_matrix_free, inverse_mass_operator_data);
+  }
+  else if(data.preconditioner == Preconditioner::AMG)
+  {
+    AMGData amg_data;
+#ifdef DEAL_II_WITH_TRILINOS
+    amg_data.ml_data.smoother_type   = "Chebyshev";
+    amg_data.ml_data.smoother_sweeps = 6;
+#endif
+    preconditioner =
+      std::make_shared<PreconditionerAMG<MassOperatorType, Number>>(mass_operator,
+                                                                    true /* initialize */,
+                                                                    amg_data);
+  }
+  else if(data.preconditioner == Preconditioner::AdditiveSchwarz)
+  {
+    preconditioner =
+      std::make_shared<AdditiveSchwarzPreconditioner<MassOperatorType>>(mass_operator,
+                                                                        true /* initialize */);
+  }
+  else
+  {
+    AssertThrow(false, dealii::ExcMessage("This type of preconditioner is not implemented."));
+  }
+
+  // Setup global CG solver.
+  Krylov::SolverCG<MassOperatorType, PreconditionerBase<Number>, VectorType> solver(
+    mass_operator, *preconditioner, data.solver_data);
+
+  dealii::Utilities::MPI::RemotePointEvaluation<dim> rpe_source(data.rpe_data);
+
+  // The sequence of integration points follows from the sequence of points as encountered during
+  // cell batch loop.
+  std::vector<dealii::Point<dim>> integration_points_target =
+    collect_integration_points<dim, n_components, Number>(target_matrix_free,
+                                                          dof_index,
+                                                          quad_index);
+
+  rpe_source.reinit(integration_points_target,
+                    source_dof_handler.get_triangulation(),
+                    *source_mapping);
+
+  if(not rpe_source.all_points_found())
+  {
+    write_points_in_dummy_triangulation(
+      integration_points_target, "./", "all_points", 0, source_dof_handler.get_communicator());
+
+    std::vector<dealii::Point<dim>> points_not_found;
+    points_not_found.reserve(integration_points_target.size());
+    for(unsigned int i = 0; i < integration_points_target.size(); ++i)
+    {
+      if(not rpe_source.point_found(i))
+      {
+        points_not_found.push_back(integration_points_target[i]);
+      }
+    }
+
+    write_points_in_dummy_triangulation(
+      points_not_found, "./", "points_not_found", 0, source_dof_handler.get_communicator());
+
+    AssertThrow(rpe_source.all_points_found(),
+                dealii::ExcMessage(
+                  "Could not interpolate source grid vector in target grid. "
+                  "Points exported to `./all_points.pvtu` and `./points_not_found.pvtu`"));
+  }
+
+  CellIntegrator<dim, n_components, Number> fe_eval(target_matrix_free, dof_index, quad_index);
```

Comment:
This object should be created in `assemble_projection_rhs()`, just as in `collect_integration_points()`, in my opinion.

Reply:
moved